lib/bulk_processor.rb

```diff
@@ -26,22 +28,40 @@ def initialize(stream:, processor_class:, payload: {})
   end
 
   # Validate the CSV and enqueue if for processing in the background.
-  def start
+  def start(file_class: S3File)
+    if file_class.new(key).exists?
+      errors << "Already processing #{key}, please wait for it to finish"
+      return false
+    end
+
+    encoded_contents = StreamEncoder.new(stream).encoded
+
     csv = ValidatedCSV.new(
-      StreamEncoder.new(stream).encoded,
+      encoded_contents,
       processor_class.required_columns,
       processor_class.optional_columns
     )
 
     if csv.valid?
-      Job.perform_later(csv.row_hashes, processor_class.name, payload)
+      perform_later(file_class, encoded_contents)
     else
-      @errors = csv.errors
+      errors.concat(csv.errors)
     end
-    @errors.empty?
+    errors.empty?
   end
 
   private
 
-  attr_reader :stream, :processor_class, :payload
+  attr_reader :key, :stream, :processor_class, :payload
+
+  def perform_later(file_class, contents)
+    file = file_class.new(key)
+    file.write(contents)
```

Comment:
This writes the file to S3, so we need to do this step here. I've separated `key` and `stream` (rather than taking just a `file`) because `ActionDispatch`'s `UploadedFile` doesn't quack like a plain ol' ruby `File`.